### PR TITLE
PE-33-ImplmentLoginPage

### PR DESCRIPTION
**Pull request by txt020**
**What I did:**
Implemented login and password widgets for login screen and signup screen. The user can now create an account and login with the same account or an existing account. Their info will be stored in Firebase.

**How to test:**

- Pull code from branch
- First thing you should see is the login screen.
 image
You can either login with an existing account or create a new one.
- Click signup to create an account then enter an email and password.
- Both screens will redirect you the a basic homes screen and in your terminal the user uid will be displayed. image
- Basic Home Screen:  
image

### CODE REVIEW

**Review by ayeejane — CHANGES_REQUESTED** [review]
A few things: 

1. When signing up, let the user know if the passwords they entered doesn't match.
2. Entering unmatched passwords still lets the user create their account. 
3. After signing up, let the user know their account has been created. Right now, after clicking `Register` nothing happens so there's no way of knowing if the user can proceed to logging in.
4. It lets the user sign in without signing up first. 
5. Reset password functionality is not working.

File: lib/login_page.dart [hunk 1]
```diff
@@ -45,26 +63,115 @@ class _LoginPageState extends State<LoginPage> {
...
+      Navigator.push(
+        context,
+        MaterialPageRoute(
+          builder: (context) {
+            return HomePage();
+          },
+        ),
+      );
+    }
+  }
```
Comment (ayeejane) [inline lib/login_page.dart]: Right now, this is just a page route. Find a way to connect it to firebase

File: lib/login_page.dart [hunk 2]
```diff
@@ -45,26 +63,115 @@ class _LoginPageState extends State<LoginPage> {
...
+                  child: Text('signup'),
+                )),
+          ],
+        ));
+  }
+
+  Widget _forgotPassword(BuildContext context) {
+    return TextButton(onPressed: null, child: Text('forgot password?'));
+  }
+
```
Comment (ayeejane) [inline lib/login_page.dart]: This button doesn't work, find a way to redirect the user to another screen so they can reset their password

File: lib/sign_in.dart [hunk 3]
```diff
@@ -5,6 +5,10 @@ import 'package:google_sign_in/google_sign_in.dart';
 final FirebaseAuth _auth = FirebaseAuth.instance;
 final GoogleSignIn googleSignIn = GoogleSignIn();
 
+Future<String> login() async {
+  //return true;
+}
+
```
Comment (ayeejane) [inline lib/sign_in.dart]: This function doesn't do anything, consider deleting it

File: lib/signup_page.dart [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                return 'Retype Password';
+              } else if (_passwordValid.hasMatch(value)) {
+                return 'Password does not match';
+              }
+              _password = value;
+              print('Confirm pass good');
+              return null;
+            },
+            textAlign: TextAlign.center,
+          ),
```
Comment (ayeejane) [inline lib/signup_page.dart]: This validator is not working properly. It doesn't correctly check if the passwords don't match

File: lib/test.dart [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                }
+              },
+              child: Text('Submit'),
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+}
```
Comment (ayeejane) [inline lib/test.dart]: What does this file do?

**Review by ayeejane — CHANGES_REQUESTED** [review]
Pleae separate your files into screens and services folder

**Comment by tonynguyen98** [comment]
I haven't tested this myself, but judging from @ayeejane's review, it seems like the majority of your task isn't working as intended. Please test your work before making a pull request.

**Review by davidsquines — CHANGES_REQUESTED** [review]
I tested the results and can validate Jane's findings:

1. Mismatching passwords still allow a new user to register
2. When entering an email, if the email already exists within firebase, the user is not notified. Please implement a notification where the user's email already exists within the system 
3. Password check functionality is not working. The app only seems to check if the email is present in the firebase database. it does not verify the password linked with the user account.

**Review by davidsquines — CHANGES_REQUESTED** [review]
After the user successfully creates the account, the user is immediately taken into the onboarding screen on the app. In other cases of register/login on other apps, the user is usually taken back to the login screen to test their credentials. Please implement this way instead

**Review by ayeejane — APPROVED** [review]
Everything works as intended now

**Review by davidsquines — APPROVED** [review]
Everything works as intended

File: lib/screens/forgot_password.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                    setState(() {
+                      _email = value;
+                    });
+                  },
+                ),
+                Padding(
+                  padding: const EdgeInsets.symmetric(vertical: 16.0),
+                  child: ElevatedButton(
+                    onPressed: () {
+                      auth.sendPasswordResetEmail(email: _email);
```
Comment (tonynguyen98) [inline lib/screens/forgot_password.dart]: lets move this to a services file for all of authentication, kinda like how you did `signInWithGoogle()`.

File: lib/screens/login_page.dart [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+  _LoginPageState createState() => _LoginPageState();
+}
+
+final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
+final RegExp _emailValid = RegExp(
+    r"^[a-zA-Z0-9.a-zA-Z0-9.!#$%&'*+-/=?^_`{|}~]+@[a-zA-Z0-9]+\.[a-zA-Z]+");
+final RegExp _passwordValid =
+    RegExp(r'^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[!@#\$&*~]).{8,}$');
+
+String _userEmail;
```
Comment (tonynguyen98) [inline lib/screens/login_page.dart]: Please move this out of the global and into `_LoginPageState`

File: lib/screens/home_page.dart [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: Container(color: Colors.orange[100]),
+    );
+  }
+}
```
Comment (tonynguyen98) [inline lib/screens/home_page.dart]: Remember, after merging with main, use the home page that @MaxLian11 had created.

File: lib/screens/login_page.dart [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    });
+  }
+
+  Future<String> signIn(_formKey) async {
+    if (_formKey.currentState.validate()) {
+      try {
+        UserCredential result = await _auth.signInWithEmailAndPassword(
+          email: email,
+          password: password,
+        );
```
Comment (tonynguyen98) [inline lib/screens/login_page.dart:81]: Send this to a services file and call it here.

File: lib/screens/signup_page.dart [hunk 10]
```diff
@@ -0,0 +1,156 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:firebase_auth/firebase_auth.dart';
+import 'package:flutter/services.dart';
+import 'package:project_eureka_flutter/services/auth.dart';
+
+import 'login_page.dart';
+
+final FirebaseAuth _auth = FirebaseAuth.instance;
+final GlobalKey<FormState> _formKey = GlobalKey<FormState>();
```
Comment (tonynguyen98) [inline lib/screens/signup_page.dart:10]: move these into `_SignupPageState()`

File: lib/screens/signup_page.dart [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+              padding: const EdgeInsets.symmetric(vertical: 16.0),
+              child: ElevatedButton(
+                onPressed: () async {
+                  print('in pressed');
+                  if (_formKey.currentState.validate()) {
+                    _formKey.currentState.save();
+                    try {
+                      final newUser =
+                          await _auth.createUserWithEmailAndPassword(
+                              email: _userEmail, password: _password);
```
Comment (tonynguyen98) [inline lib/screens/signup_page.dart:114]: services file

**Review by tonynguyen98 — CHANGES_REQUESTED** [review]
Please make an auth services file and move anything that has to do with firebase into there. You've also got a lot of repeating code, but since it'll be a new issue to implement the UI for this, the person who's building the UI will clean up the code.

**Comment by tonynguyen98** [comment]
> Please make an auth services file and move anything that has to do with firebase into there. You've also got a lot of repeating code, but since it'll be a new issue to implement the UI for this, the person who's building the UI will clean up the code.

Ignore this message, just merge this code without making a services file. I will do that myself as my next task. This task itself already have been in-progress for 2 months.

**Review by tonynguyen98 — APPROVED** [review]

**Pull request merged**